# Alex's data quest

In [132]:
import pandas as pd

In [133]:
#data = pd.read_csv('data.csv', error_bad_lines=False, encoding="utf-8") # CSV old way
data = pd.read_json("data_5scheduler.json") # json new way

Look at the data:

In [134]:
print(data.describe(include="object"))

                title   identifier description  source instructors offered  \
count            4443         4443        4443    4443        4443    4443   
unique           3761         4234        3989       5        1231     139   
top     Senior Thesis  PHYS-178-KS              Pomona          []           
freq               56            2          95    1446        2023    1559   

       prerequisites corequisites  
count           4443         4443  
unique           716           28  
top                                
freq            3359         4411  


### Duplicates:
It looks like there is only 3989 unique course descripition so let's remove duplicates based on 'description' column.
There are also rows with empty descriptions, which are not helpful

In [135]:
print(len(data))
data = data.drop_duplicates(subset='description')
data = data[data["description"] != ""]
print(len(data))

4443
3988


In [136]:
data.head(10)

title   identifier  \
0                  Introduction to American Cultures  AMST-103-HM   
1                         Print and American Culture  AMST-115-HM   
2                               Hyphenated Americans  AMST-120-HM   
3    Life: Knowledge, Belief, and Cultural Practices  ANTH-110-HM   
4  Introduction to the Anthropology of Science an...  ANTH-111-HM   
5                                   War and Conflict  ANTH-115-HM   
6                                      Rationalities  ANTH-134-HM   
7                 A History of Landscape Photography  ARHI-131-HM   
8              Modern and Contemporary Art Practices   ART-002-HM   
9                                        Photography   ART-033-HM   

                                         description      source  credits  \
0  An interdisciplinary introduction to principal...  HarveyMudd      300   
1  Covers numerous developments in American print...  HarveyMudd      300   
2  A focus on the experience of immigrants in the...  HarveyMudd      300   
3  An exploration of cultural attitudes toward li...  HarveyMudd      300   
4  An introduction to science and technology as c...  HarveyMudd      300   
5  “The wings of the butterfly—that cause the hur...  HarveyMudd      300   
6  What does it mean to be rational? Does it mean...  HarveyMudd      300   
7  This course explores how photographic landscap...  HarveyMudd      300   
8  This class is an experimental lecture style ar...  HarveyMudd      300   
9  Approaching the medium from an artistic perspe...  HarveyMudd      300   

          instructors                  offered  \
0             [Staff]                            
1        [Anup Gampa]                            
2          [Balseiro]                            
3           [de Laet]                            
4  [Marianne De Laet]                            
5           [de Laet]                            
6           [de Laet]  Offered alternate years   
7           [Fandell]                            
8           [Fandell]                            
9           [Fandell]                            

                                       prerequisites corequisites  \
0                                                                   
1                                                                   
2                                                                   
3                                                                   
4                                                                   
5                                                                   
6  Any introductory course in anthropology or any...                
7                                                                   
8                                                                   
9                                         ART002 HM                 

   currently_offered  fee  
0              False    0  
1               True    0  
2              False    0  
3              False    0  
4               True    0  
5              False    0  
6              False    0  
7              False    0  
8              False    0  
9              False  150

### tf-idf with scikit-learn
[Description](https://monkeylearn.com/blog/what-is-tf-idf/)

[Usage](https://kavita-ganesan.com/tfidftransformer-tfidfvectorizer-usage-differences/#.Y1M42ezMJhF)

Here is an example of how TfIdf would work if our documents were the following 4 sentences:

In [137]:
import sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]
vectorizer = TfidfVectorizer(use_idf=True)
vectors = vectorizer.fit_transform(corpus)
firstv = vectors[0]
df = pd.DataFrame(firstv.T.todense(), index=vectorizer.get_feature_names(), columns=["tfidf"])
df = df.sort_values(by=["tfidf"], ascending = False)
print("TfIdf values for the first sentence")
print(df)


TfIdf values for the first sentence
             tfidf
first     0.580286
document  0.469791
is        0.384085
the       0.384085
this      0.384085
and       0.000000
one       0.000000
second    0.000000
third     0.000000


/Users/alex/Library/Python/3.9/lib/python/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In the example above we can see the importance of each word ranked for the first sentence `'this is the first document'`. So, for example the word `first` is important since it doesn't appear in any other document. The word `the` is not as important since it appears in all other documents. And the word `third` is not important at all since it doesn't even appear in the first document.

Function `tfidf(word, data)` takes in the word we are interested in and the data we are looking at. The function returns an updated dataframe with a new column `"score"` that gives each class a score of importance based on the input word. 

In [138]:
def tfidf(word, data):
    print(data.loc[0, "description"])
    corpus = list(data.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)

    score_for_word = []
    words = vectorizer.get_feature_names()
    try:
        index = words.index(word)
    except:
        print("'" + word + "'" + " is not mentioned in any course descriptions")
        return

    for i in range(0, len(corpus)):
        value = vectors[i].T.todense()[index]
        score_for_word.append(value)

    score_for_word = [float(i) for i in score_for_word] # type cast each score to a float

    data["score"] = score_for_word
    data = data.sort_values(by=["score"], ascending = False)
    return data

For example, let's say we are interested in ranking all of the classes based on the word `computer`:

In [139]:
tfidf('computer', data).head(20)

An interdisciplinary introduction to principal themes in American culture taught by an intercollegiate faculty team.


/Users/alex/Library/Python/3.9/lib/python/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


title     identifier  \
132                          Computer Science Seminar    CSCI-181-HM   
525                Special Topics in Computer Science    CSCI-181-CM   
1490                      Computer Science Colloquium    CSCI-188-PO   
426        Introduction to Computational Neuroscience   BIOL-133L-KS   
1491                  Computer Science Senior Seminar    CSCI-190-PO   
1060            Computational Physics and Engineering    PHYS-100-KS   
518                  Fundamentals of Computer Science    CSCI-052-CM   
3358            Computational Physics and Engineering    PHYS-100-KS   
137                       Computer Science Colloquium    CSCI-195-HM   
106      Introduction to Biology and Computer Science  CSCI-005GR-HM   
515                    Data Structures and Algorithms    CSCI-046-CM   
286                    The Harmony of Sound and Light      MS-127-HM   
157      Digital Electronics and Computer Engineering    ENGR-085-HM   
54                     Advanced Computational Biology    BIOL-188-HM   
105                  Introduction to Computer Science    CSCI-005-HM   
135   Computer Science Research and Independent Study    CSCI-186-HM   
108       Principles and Practice of Computer Science    CSCI-042-HM   
1493                                   Senior Project    CSCI-192-PO   
116                                 Computer Networks    CSCI-125-HM   
517                  Introduction to Computer Science    CSCI-051-CM   

                                            description            source  \
132   Advanced topics of current interest in compute...        HarveyMudd   
525   Selected topics in computer science. May be re...  ClaremontMckenna   
1490  Colloquium presentations and discussions of to...            Pomona   
426   This course provides computational skills for ...  ClaremontMckenna   
1491  Reading, discussion and presentation of resear...            Pomona   
1060  This course is a comprehensive introduction to...  ClaremontMckenna   
518   A solid foundation in functional programming, ...  ClaremontMckenna   
3358  This course is a comprehensive introduction to...           Scripps   
137   Oral presentations and discussions of selected...        HarveyMudd   
106   This course introduces fundamental concepts fr...        HarveyMudd   
515   An introduction to abstract data types and fun...  ClaremontMckenna   
286   New technology has created exciting new opport...        HarveyMudd   
157   This course provides an introduction to elemen...        HarveyMudd   
54    Computational algorithms and methods used in t...        HarveyMudd   
105   Introduction to elements of computer science. ...        HarveyMudd   
135   A research or development project under comput...        HarveyMudd   
108   Accelerated breadth-first introduction to comp...        HarveyMudd   
1493  A substantial, independent or group project in...            Pomona   
116   Principles and analysis techniques for interne...        HarveyMudd   
517   First computer science course intended for stu...  ClaremontMckenna   

      credits                                      instructors  \
132         0                                          [Staff]   
525       100                                               []   
1490        0                                [Joseph C Osborn]   
426       100                                               []   
1491       25                                [Joseph C Osborn]   
1060      100                                     [Scot Gould]   
518       100                                               []   
3358      100                                               []   
137        50                             [Melissa E. O'Neill]   
106       300                             [Wu, Bush (Biology)]   
515       100                                     [Mark Huber]   
286       300                                          [Alves]   
157       300                             [David Money Harr

These are the first 20 instances of the classes that are most related to the word `computer` ranked in descending order. So, we could recomend a student who is interested in `computer` to take these classes.